sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala

```diff
@@ -888,6 +889,179 @@ class StringExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
+        .filter(_.nonEmpty).foreach { format =>
+        checkExceptionInExpression[IllegalArgumentException](
+          ToNumber(Literal("454"), Literal(format)), s"Format '$format' used for" +
+            " parsing string to number or formatting number to string is invalid")
+      }
+
+      val format1 = 0.until(input.length).map(_ => '0').mkString
+      val format2 = 0.until(input.length).map(_ => '9').mkString
+      val format3 = 0.until(input.length + 1).map(_ => '0').mkString
+      val format4 = 0.until(input.length + 1).map(_ => '9').mkString
+      val format5 = 0.until(input.length + 2).map(_ => '0').mkString
+      val format6 = 0.until(input.length + 2).map(_ => '9').mkString
+      Seq(format1, format2, format3, format4, format5, format6).foreach { format =>
+        checkEvaluation(ToNumber(Literal(input), Literal(format)), Decimal(input))
+      }
+    }
+
+    // Test '.' and 'D'
+    checkExceptionInExpression[IllegalArgumentException](
+      ToNumber(Literal("454.2"), Literal("999")),
+      "Format '999' used for parsing string to number or formatting number to string is invalid")
+    checkExceptionInExpression[IllegalArgumentException](
+      ToNumber(Literal("454.23"), Literal("999.9")),
+      "Format '999.9' used for parsing string to number or formatting number to string is invalid")
+    checkExceptionInExpression[IllegalArgumentException](
+      ToNumber(Literal("454.23"), Literal("000.0")),
+      "Format '000.0' used for parsing string to number or formatting number to string is invalid")
+    checkExceptionInExpression[IllegalArgumentException](
+      ToNumber(Literal("454.23"), Literal("000D0")),
+      "Format '000D0' used for parsing string to number or formatting number to string is invalid")
+    checkExceptionInExpression[IllegalArgumentException](
+      ToNumber(Literal("454.23"), Literal("00.00")),
+      "Format '00.00' used for parsing string to number or formatting number to string is invalid")
+    checkExceptionInExpression[IllegalArgumentException](
+      ToNumber(Literal("454.23"), Literal("00D00")),
+      "Format '00D00' used for parsing string to number or formatting number to string is invalid")
+    checkExceptionInExpression[IllegalArgumentException](
+      ToNumber(Literal("454.23"), Literal("0000.0")),
+      "Format '0000.0' used for parsing string to number or formatting number to string is invalid")
+    checkExceptionInExpression[IllegalArgumentException](
+      ToNumber(Literal("454.23"), Literal("0000D0")),
+      "Format '0000D0' used for parsing string to number or formatting number to string is invalid")
+    checkExceptionInExpression[IllegalArgumentException](
+      ToNumber(Literal("454.23"), Literal("00.000")),
+      "Format '00.000' used for parsing string to number or formatting number to string is invalid")
+    checkExceptionInExpression[IllegalArgumentException](
+      ToNumber(Literal("454.23"), Literal("00D000")),
```

Comment:
if you really want to test all the cases, please write the test cases only with dot, and run it again after replacing dot with `D`.